# Yieldigo Analysis

### Assumptions:
 - **Zones**
     - Shops in one zone must have the same prices
 - **Sites**
     - Same price products sorted into price groups via ZoneId
     - Sorted by region (I'll consider it SiteName) and shop type (zone)
 - **Articles**
     - Product list
     - ArticleId key
 -  **Sales**
     -  Daily agg
     -  product number, day of sale, site, cost price, sales price - eliminate errors
     -  prices equals to amount 1

In [195]:
import pandas as pd
import numpy as np

## Loading data

In [176]:
zones = pd.read_csv("data/zones.csv")
zones

ZoneId     ZoneName
0       1  Hypermarket
1       2  Supermarket

In [177]:
sites = pd.read_csv("data/sites.csv")
sites

SiteId  ZoneId   SiteName
0       4       1  Londýnská
1       5       1   Belgická
2       6       1   Vídeňská
3       7       2    Pražská
4       8       2   Plzeňská

In [196]:
articles = pd.read_csv("data/articles.csv")
articles

ArticleId                     ArticleName
0      170295                   BERLINKI 250G
1      172551       BOŽKOV SPE.KÁVOVÝ 30%0,5L
2      172671        GNOCCHI 40% 500G SPINACI
3      167108      GRANINI NEKT.1l HRUŠKA 50%
4      176139                HRANOLKY B" 1KG"
5      170069     JACOBS PORCE CAF.LATTÉ 250G
6      174022          JÄGERMEISTER 35%  0,7L
7      165811      JČ SMETANA KE ŠL.33% 250ML
8      165954     LIPÁNEK UHT MLÉKO 250ML KAK
9      169281  MAGGI ASIA 118G INDIE SMAŽ.NUD
10     199219       NORDIC ICE VODKA37,5%0,5L
11     171979        POMAZÁNKA 120G VAJÍČKOVÁ
12     194516   POMAZÁNKA CAMEMBERTOVÁ 3X120G
13     200838         PRAŽSKÁ VODKA 37,5%0,5L
14     170936          ŠPENÁTOVÝ PROTLAK 400G

In [197]:
sales = pd.read_csv("data/sales.csv")
sales

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice
0         170936       4       1  2019-05-09       7.0   39.0      30.15
1         171979       7       2  2019-01-18       4.0   43.5      28.11
2         200838       5       1  2019-05-16      10.0   95.0     159.49
3         176139       7       2  2018-12-21       5.0  199.0     147.97
4         172551       5       1  2019-04-11       1.0   86.7      80.76
...          ...     ...     ...         ...       ...    ...        ...
17992     170936       6       1  2018-09-10       5.0   39.0      30.06
17993     199219       6       1  2019-01-23       4.0   89.0      77.90
17994     165811       5       1  2018-12-13       4.0   21.9      15.74
17995     167108       8       2  2019-04-01       1.0   27.9      18.62
17996     170936       8       2  2018-12-05      11.0   39.0      30.08

[17997 rows x 7 columns]

#### Notes:
**The data load displays sample of the data. I loaded the data succesfully, now I'm checking for general issues, missing values, data types. I would like to expand sales table with other tables to create one table and perform analysis.**

In [219]:
def missing_values(df):
    for i in df.columns:
        pct_missing = np.mean(df[i].isnull()) * 100
        print("{}-{}%".format(i, pct_missing))

In [220]:
def value_count(df):
    for i in df.columns:
        print(df[i].value_counts())
        print("\n")

In [221]:
missing_values(zones)

ZoneId-0.0%
ZoneName-0.0%


In [222]:
missing_values(sites)

SiteId-0.0%
ZoneId-0.0%
SiteName-0.0%


In [223]:
missing_values(articles)

ArticleId-0.0%
ArticleName-0.0%


In [224]:
missing_values(sales)

ArticleId-0.0%
SiteId-0.0%
ZoneId-0.0%
Date-0.0%
Quantity-0.0%
Price-0.07223426126576653%
CostPrice-0.15002500416736123%


In [225]:
value_count(zones)

ZoneId
1    1
2    1
Name: count, dtype: int64


ZoneName
Hypermarket    1
Supermarket    1
Name: count, dtype: int64




In [226]:
value_count(sites)

SiteId
4    1
5    1
6    1
7    1
8    1
Name: count, dtype: int64


ZoneId
1    3
2    2
Name: count, dtype: int64


SiteName
Londýnská    1
Belgická     1
Vídeňská     1
Pražská      1
Plzeňská     1
Name: count, dtype: int64




In [227]:
value_count(articles)

ArticleId
170295    1
172551    1
172671    1
167108    1
176139    1
170069    1
174022    1
165811    1
165954    1
169281    1
199219    1
171979    1
194516    1
200838    1
170936    1
Name: count, dtype: int64


ArticleName
BERLINKI 250G                     1
BOŽKOV SPE.KÁVOVÝ 30%0,5L         1
GNOCCHI 40% 500G SPINACI          1
GRANINI NEKT.1l HRUŠKA 50%        1
HRANOLKY B" 1KG"                  1
JACOBS PORCE CAF.LATTÉ 250G       1
JÄGERMEISTER 35%  0,7L            1
JČ SMETANA KE ŠL.33% 250ML        1
LIPÁNEK UHT MLÉKO 250ML KAK       1
MAGGI ASIA 118G INDIE SMAŽ.NUD    1
NORDIC ICE VODKA37,5%0,5L         1
POMAZÁNKA 120G VAJÍČKOVÁ          1
POMAZÁNKA CAMEMBERTOVÁ 3X120G     1
PRAŽSKÁ VODKA 37,5%0,5L           1
ŠPENÁTOVÝ PROTLAK 400G            1
Name: count, dtype: int64




In [228]:
value_count(sales)

ArticleId
170936    1853
174022    1801
200838    1680
172551    1640
167108    1624
199219    1604
165811    1560
170295    1185
176139    1074
171979    1012
194516     966
170069     656
169281     537
172671     503
165954     302
Name: count, dtype: int64


SiteId
4    3741
5    3720
8    3461
6    3443
7    3379
9     253
Name: count, dtype: int64


ZoneId
1    10904
2     7093
Name: count, dtype: int64


Date
2018-11-15    69
2018-11-16    68
2018-11-14    68
2018-11-23    67
2018-11-24    65
              ..
2018-12-06    33
2018-12-10    33
2018-08-06    31
2018-07-30    28
2018-12-24    21
Name: count, Length: 389, dtype: int64


Quantity
 1.0     3029
 2.0     2282
 3.0     1720
 4.0     1496
 5.0     1212
         ... 
-42.0       1
 62.0       1
 56.0       1
 76.0       1
-25.0       1
Name: count, Length: 102, dtype: int64


Price
39.00     1676
21.90     1541
95.00     1184
89.00     1098
239.00     941
          ... 
88.54        1
38.48        1
42.12        1
21.10  

In [229]:
zones.dtypes

ZoneId       int64
ZoneName    object
dtype: object

In [230]:
sites.dtypes

SiteId       int64
ZoneId       int64
SiteName    object
dtype: object

In [231]:
articles.dtypes

ArticleId       int64
ArticleName    object
dtype: object

In [232]:
sales.dtypes

ArticleId      int64
SiteId         int64
ZoneId         int64
Date          object
Quantity     float64
Price        float64
CostPrice    float64
dtype: object

#### Notes:
**What I can read from the dtypes, missing_values and value_counts?**
- Zones, Sites, and Articles data are clean, have the correct data type with expected value counts, and are ready to merge.
- Sales data contain missing values within price and cost_price.
- Date should be in date-time format
- Value counts showed some inconsistencies (quantity has values below zero, sites have id=9, which does not exist in the sites table, and we have only 5 stores).

**I have no missing values in primary keys, therefore, I'm able to merge and then address these issues (following the assumptions).**

In [233]:
df = sales.merge(zones, on="ZoneId", how="inner")
df

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0      30.15   
1         171979       7       2  2019-01-18       4.0   43.5      28.11   
2         200838       5       1  2019-05-16      10.0   95.0     159.49   
3         176139       7       2  2018-12-21       5.0  199.0     147.97   
4         172551       5       1  2019-04-11       1.0   86.7      80.76   
...          ...     ...     ...         ...       ...    ...        ...   
17992     170936       6       1  2018-09-10       5.0   39.0      30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.90   
17994     165811       5       1  2018-12-13       4.0   21.9      15.74   
17995     167108       8       2  2019-04-01       1.0   27.9      18.62   
17996     170936       8       2  2018-12-05      11.0   39.0      30.08   

          ZoneName  
0      Hypermarket  
1      Supermarket  
2      Hypermarket  
3      Supermarket  
4      Hypermarket  
...            ...  
17992  Hypermarket  
17993  Hypermarket  
17994  Hypermarket  
17995  Supermarket  
17996  Supermarket  

[17997 rows x 8 columns]

In [234]:
df = df.merge(sites[["SiteId","SiteName"]], on="SiteId", how="left")
df

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0      30.15   
1         171979       7       2  2019-01-18       4.0   43.5      28.11   
2         200838       5       1  2019-05-16      10.0   95.0     159.49   
3         176139       7       2  2018-12-21       5.0  199.0     147.97   
4         172551       5       1  2019-04-11       1.0   86.7      80.76   
...          ...     ...     ...         ...       ...    ...        ...   
17992     170936       6       1  2018-09-10       5.0   39.0      30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.90   
17994     165811       5       1  2018-12-13       4.0   21.9      15.74   
17995     167108       8       2  2019-04-01       1.0   27.9      18.62   
17996     170936       8       2  2018-12-05      11.0   39.0      30.08   

          ZoneName   SiteName  
0      Hypermarket  Londýnská  
1      Supermarket    Pražská  
2      Hypermarket   Belgická  
3      Supermarket    Pražská  
4      Hypermarket   Belgická  
...            ...        ...  
17992  Hypermarket   Vídeňská  
17993  Hypermarket   Vídeňská  
17994  Hypermarket   Belgická  
17995  Supermarket   Plzeňská  
17996  Supermarket   Plzeňská  

[17997 rows x 9 columns]

In [235]:
missing_values(df)

ArticleId-0.0%
SiteId-0.0%
ZoneId-0.0%
Date-0.0%
Quantity-0.0%
Price-0.07223426126576653%
CostPrice-0.15002500416736123%
ZoneName-0.0%
SiteName-1.4057898538645328%


In [236]:
test = df.copy()
test = test[test["SiteName"].isna() == True]
test
# ---------- I can either dropna() or while merging sites with df, select how="inner" for resulting df with 17744 -------

ArticleId  SiteId  ZoneId        Date  Quantity   Price  CostPrice  \
53        194516       9       2  2018-11-26      11.0   40.64      27.73   
164       200838       9       2  2018-11-13       5.0   99.75      87.24   
373       200838       9       2  2018-11-27       1.0   99.75      87.34   
385       172551       9       2  2018-11-21       1.0   95.03      84.17   
411       165811       9       2  2018-11-17       5.0   23.00      15.68   
...          ...     ...     ...         ...       ...     ...        ...   
17793     171979       9       2  2018-11-16       2.0   44.73      28.96   
17830     170936       9       2  2018-11-24      14.0   40.95      30.66   
17841     167108       9       2  2018-11-01       2.0   26.15      19.05   
17896     176139       9       2  2018-11-06       3.0  240.45     152.06   
17950     167108       9       2  2018-11-15       1.0   40.22      18.99   

          ZoneName SiteName  
53     Supermarket      NaN  
164    Supermarket      NaN  
373    Supermarket      NaN  
385    Supermarket      NaN  
411    Supermarket      NaN  
...            ...      ...  
17793  Supermarket      NaN  
17830  Supermarket      NaN  
17841  Supermarket      NaN  
17896  Supermarket      NaN  
17950  Supermarket      NaN  

[253 rows x 9 columns]

In [237]:
df = df.merge(articles, on="ArticleId", how="inner")
df

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0      30.15   
1         171979       7       2  2019-01-18       4.0   43.5      28.11   
2         200838       5       1  2019-05-16      10.0   95.0     159.49   
3         176139       7       2  2018-12-21       5.0  199.0     147.97   
4         172551       5       1  2019-04-11       1.0   86.7      80.76   
...          ...     ...     ...         ...       ...    ...        ...   
17992     170936       6       1  2018-09-10       5.0   39.0      30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.90   
17994     165811       5       1  2018-12-13       4.0   21.9      15.74   
17995     167108       8       2  2019-04-01       1.0   27.9      18.62   
17996     170936       8       2  2018-12-05      11.0   39.0      30.08   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17995  Supermarket   Plzeňská  GRANINI NEKT.1l HRUŠKA 50%  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17997 rows x 10 columns]

In [238]:
df[["Quantity", "Price", "CostPrice"]].describe()

Quantity         Price     CostPrice
count  17997.000000  17984.000000  17970.000000
mean       6.260599     81.778959     64.442784
std        8.413260     72.251808     58.840182
min      -42.000000      8.700000      5.830000
25%        2.000000     26.900000     18.580000
50%        4.000000     46.900000     30.130000
75%        9.000000     95.000000     82.480000
max      432.000000    399.000000    383.010000

#### Notes
**After merge:**
- SiteName now has 1.40% missing values, a small number totaling 253 rows. I have the option to simply drop rows with missing values. In a real-life scenario, there is also the option to investigate the missing SiteId and input the correct value or use web scraping (for an external data source).
- Personally, I think the rest of the data is relatively clean, and instead of dropping the missing values, I’m going to impute them with a dummy variable. This way, we can still extract insights from the 1.4% for the rest of the dataset.

## Cleaning of DF
**Steps:**
- Impute SiteName NaN values with "Testingová."
- Fill CostPrice and Price based on the price of ZoneId.
- Change Date format to the correct type (possibly also Quantity).
- Check prices within one shop.
- Check for quantity values below zero.
- Remove duplicates.
- Check for the same purchase prices for one product within the same store.

#### Imput Value - "Testingová"

In [135]:
missing_name = df.copy()
missing_name = missing_name.fillna("Testingová")
missing_name

ArticleId  SiteId  ZoneId        Date  Quantity  Price CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0     30.15   
1         171979       7       2  2019-01-18       4.0   43.5     28.11   
2         200838       5       1  2019-05-16      10.0   95.0    159.49   
3         176139       7       2  2018-12-21       5.0  199.0    147.97   
4         172551       5       1  2019-04-11       1.0   86.7     80.76   
...          ...     ...     ...         ...       ...    ...       ...   
17992     170936       6       1  2018-09-10       5.0   39.0     30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.9   
17994     165811       5       1  2018-12-13       4.0   21.9     15.74   
17995     167108       8       2  2019-04-01       1.0   27.9     18.62   
17996     170936       8       2  2018-12-05      11.0   39.0     30.08   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17995  Supermarket   Plzeňská  GRANINI NEKT.1l HRUŠKA 50%  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17997 rows x 10 columns]

In [136]:
missing_name_check = missing_name[missing_name["SiteName"].isna() == True]
missing_name_check

Empty DataFrame
Columns: [ArticleId, SiteId, ZoneId, Date, Quantity, Price, CostPrice, ZoneName, SiteName, ArticleName]
Index: []

In [137]:
df["SiteName"] = missing_name["SiteName"]

In [138]:
df

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0      30.15   
1         171979       7       2  2019-01-18       4.0   43.5      28.11   
2         200838       5       1  2019-05-16      10.0   95.0     159.49   
3         176139       7       2  2018-12-21       5.0  199.0     147.97   
4         172551       5       1  2019-04-11       1.0   86.7      80.76   
...          ...     ...     ...         ...       ...    ...        ...   
17992     170936       6       1  2018-09-10       5.0   39.0      30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.90   
17994     165811       5       1  2018-12-13       4.0   21.9      15.74   
17995     167108       8       2  2019-04-01       1.0   27.9      18.62   
17996     170936       8       2  2018-12-05      11.0   39.0      30.08   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17995  Supermarket   Plzeňská  GRANINI NEKT.1l HRUŠKA 50%  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17997 rows x 10 columns]

#### Remove duplicates

In [139]:
duplicates_with_first = df[df.duplicated(keep=False)]
duplicates_with_first = duplicates_with_first.sort_values(by=df.columns.tolist())
duplicates_with_first

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
2665      165811       8       2  2019-04-03       1.0   21.9      15.81   
3125      165811       8       2  2019-04-03       1.0   21.9      15.81   
15539     165811       8       2  2019-04-12      16.0   21.9      15.83   
17018     165811       8       2  2019-04-12      16.0   21.9      15.83   
12371     165811       8       2  2019-04-13       1.0   21.9      15.83   
...          ...     ...     ...         ...       ...    ...        ...   
11806     200838       8       2  2019-04-28       7.0   95.0      81.89   
6017      200838       8       2  2019-04-29       4.0   95.0      81.89   
17582     200838       8       2  2019-04-29       4.0   95.0      81.89   
3612      200838       8       2  2019-04-30      13.0   95.0      81.89   
7250      200838       8       2  2019-04-30      13.0   95.0      81.89   

          ZoneName  SiteName                 ArticleName  
2665   Supermarket  Plzeňská  JČ SMETANA KE ŠL.33% 250ML  
3125   Supermarket  Plzeňská  JČ SMETANA KE ŠL.33% 250ML  
15539  Supermarket  Plzeňská  JČ SMETANA KE ŠL.33% 250ML  
17018  Supermarket  Plzeňská  JČ SMETANA KE ŠL.33% 250ML  
12371  Supermarket  Plzeňská  JČ SMETANA KE ŠL.33% 250ML  
...            ...       ...                         ...  
11806  Supermarket  Plzeňská     PRAŽSKÁ VODKA 37,5%0,5L  
6017   Supermarket  Plzeňská     PRAŽSKÁ VODKA 37,5%0,5L  
17582  Supermarket  Plzeňská     PRAŽSKÁ VODKA 37,5%0,5L  
3612   Supermarket  Plzeňská     PRAŽSKÁ VODKA 37,5%0,5L  
7250   Supermarket  Plzeňská     PRAŽSKÁ VODKA 37,5%0,5L  

[492 rows x 10 columns]

In [140]:
df.drop_duplicates(inplace=True)
df

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
0         170936       4       1  2019-05-09       7.0   39.0      30.15   
1         171979       7       2  2019-01-18       4.0   43.5      28.11   
2         200838       5       1  2019-05-16      10.0   95.0     159.49   
3         176139       7       2  2018-12-21       5.0  199.0     147.97   
4         172551       5       1  2019-04-11       1.0   86.7      80.76   
...          ...     ...     ...         ...       ...    ...        ...   
17991     199219       5       1  2019-05-04      17.0   89.0      78.22   
17992     170936       6       1  2018-09-10       5.0   39.0      30.06   
17993     199219       6       1  2019-01-23       4.0   89.0      77.90   
17994     165811       5       1  2018-12-13       4.0   21.9      15.74   
17996     170936       8       2  2018-12-05      11.0   39.0      30.08   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17751 rows x 10 columns]

#### Notes
- Removing duplicates to reduce noise

#### Missing Value - Price
- Impute most frequent value from the same shop

In [141]:
missing_values(df)

ArticleId-0.0%
SiteId-0.0%
ZoneId-0.0%
Date-0.0%
Quantity-0.0%
Price-0.07323531068672187%
CostPrice-0.1521041068108839%
ZoneName-0.0%
SiteName-0.0%
ArticleName-0.0%


In [142]:
missing_price = df.copy()
missing_price = missing_price[missing_price["Price"].isna()]
missing_price

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
549       170069       4       1  2019-04-13       1.0    NaN      42.91   
811       172671       4       1  2019-04-05       1.0    NaN      10.91   
4521      167108       4       1  2019-04-15       1.0    NaN      18.62   
4748      176139       4       1  2019-04-01       1.0    NaN     142.00   
5043      194516       4       1  2019-04-15       1.0    NaN      27.48   
7992      176139       4       1  2019-04-11       1.0    NaN     142.00   
10294     172551       4       1  2019-04-09       1.0    NaN      79.16   
10335     171979       4       1  2019-04-01       1.0    NaN      28.82   
10367     165954       4       1  2019-04-01       1.0    NaN       5.87   
10568     172671       4       1  2019-04-09       1.0    NaN      10.91   
11157     170295       4       1  2019-04-04       1.0    NaN      15.79   
13918     194516       4       1  2019-04-11       1.0    NaN      27.51   
16599     169281       4       1  2019-04-07       1.0    NaN      15.56   

          ZoneName   SiteName                     ArticleName  
549    Hypermarket  Londýnská     JACOBS PORCE CAF.LATTÉ 250G  
811    Hypermarket  Londýnská        GNOCCHI 40% 500G SPINACI  
4521   Hypermarket  Londýnská      GRANINI NEKT.1l HRUŠKA 50%  
4748   Hypermarket  Londýnská                HRANOLKY B" 1KG"  
5043   Hypermarket  Londýnská   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
7992   Hypermarket  Londýnská                HRANOLKY B" 1KG"  
10294  Hypermarket  Londýnská       BOŽKOV SPE.KÁVOVÝ 30%0,5L  
10335  Hypermarket  Londýnská        POMAZÁNKA 120G VAJÍČKOVÁ  
10367  Hypermarket  Londýnská     LIPÁNEK UHT MLÉKO 250ML KAK  
10568  Hypermarket  Londýnská        GNOCCHI 40% 500G SPINACI  
11157  Hypermarket  Londýnská                   BERLINKI 250G  
13918  Hypermarket  Londýnská   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
16599  Hypermarket  Londýnská  MAGGI ASIA 118G INDIE SMAŽ.NUD

In [143]:
londynska = df[df["SiteName"] == "Londýnská"]
frequent_londynska = londynska.groupby("ArticleName")["Price"].agg(lambda x: x.mode()[0] if not x.mode().empty else None)
df.loc[(df["SiteName"] == "Londýnská") & (df["Price"].isna()), "Price"] = df["ArticleName"].map(frequent_londynska)

In [144]:
missing_values(df)

ArticleId-0.0%
SiteId-0.0%
ZoneId-0.0%
Date-0.0%
Quantity-0.0%
Price-0.0%
CostPrice-0.1521041068108839%
ZoneName-0.0%
SiteName-0.0%
ArticleName-0.0%


#### Missing Cost Price
- Impute same value for cost price from the same shop

In [145]:
missing_cost = df.copy()
missing_cost = missing_cost[missing_cost["CostPrice"].isna()]
missing_cost

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
603       171979       5       1  2019-03-12       1.0   48.6        NaN   
1279      167108       5       1  2019-03-05       1.0   27.9        NaN   
2601      167108       5       1  2019-03-02       1.0   27.9        NaN   
2918      172671       5       1  2019-03-07       1.0   19.9        NaN   
3019      172671       5       1  2019-03-11       1.0   19.9        NaN   
5276      194516       5       1  2019-03-10       1.0   46.5        NaN   
5693      176139       5       1  2019-03-12       1.0  197.0        NaN   
6648      172551       5       1  2019-03-01       1.0   98.9        NaN   
6733      171979       5       1  2019-03-06       1.0   48.6        NaN   
7726      165954       5       1  2019-03-03       1.0    8.7        NaN   
8327      194516       5       1  2019-03-05       1.0   46.5        NaN   
8607      194516       5       1  2019-03-14       1.0   46.5        NaN   
10817     172671       5       1  2019-03-15       1.0   19.9        NaN   
11872     174022       5       1  2019-03-04       1.0  266.0        NaN   
12827     199219       5       1  2019-03-05       1.0   89.0        NaN   
13104     169281       5       1  2019-03-14       1.0   21.9        NaN   
13858     176139       5       1  2019-03-02       1.0  197.0        NaN   
14296     171979       5       1  2019-03-15       1.0   48.6        NaN   
14433     172671       5       1  2019-03-04       1.0   19.9        NaN   
15591     199219       5       1  2019-03-02       1.0   89.0        NaN   
16265     176139       5       1  2019-03-05       1.0  197.0        NaN   
16677     194516       5       1  2019-03-03       1.0   46.9        NaN   
16924     171979       5       1  2019-03-03       1.0   43.5        NaN   
16947     169281       5       1  2019-03-08       1.0   21.9        NaN   
17048     176139       5       1  2019-03-09       1.0  197.0        NaN   
17251     167108       5       1  2019-03-04       1.0   27.9        NaN   
17900     167108       5       1  2019-03-15       1.0   27.9        NaN   

          ZoneName  SiteName                     ArticleName  
603    Hypermarket  Belgická        POMAZÁNKA 120G VAJÍČKOVÁ  
1279   Hypermarket  Belgická      GRANINI NEKT.1l HRUŠKA 50%  
2601   Hypermarket  Belgická      GRANINI NEKT.1l HRUŠKA 50%  
2918   Hypermarket  Belgická        GNOCCHI 40% 500G SPINACI  
3019   Hypermarket  Belgická        GNOCCHI 40% 500G SPINACI  
5276   Hypermarket  Belgická   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
5693   Hypermarket  Belgická                HRANOLKY B" 1KG"  
6648   Hypermarket  Belgická       BOŽKOV SPE.KÁVOVÝ 30%0,5L  
6733   Hypermarket  Belgická        POMAZÁNKA 120G VAJÍČKOVÁ  
7726   Hypermarket  Belgická     LIPÁNEK UHT MLÉKO 250ML KAK  
8327   Hypermarket  Belgická   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
8607   Hypermarket  Belgická   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
10817  Hypermarket  Belgická        GNOCCHI 40% 500G SPINACI  
11872  Hypermarket  Belgická          JÄGERMEISTER 35%  0,7L  
12827  Hypermarket  Belgická       NORDIC ICE VODKA37,5%0,5L  
13104  Hypermarket  Belgická  MAGGI ASIA 118G INDIE SMAŽ.NUD  
13858  Hypermarket  Belgická                HRANOLKY B" 1KG"  
14296  Hypermarket  Belgická        POMAZÁNKA 120G VAJÍČKOVÁ  
14433  Hypermarket  Belgická        GNOCCHI 40% 500G SPINACI  
15591  Hypermarket  Belgická       NORDIC ICE VODKA37,5%0,5L  
16265  Hypermarket  Belgická                HRANOLKY B" 1KG"  
16677  Hypermarket  Belgická   POMAZÁNKA CAMEMBERTOVÁ 3X120G  
16924  Hypermarket  Belgická        POMAZÁNKA 120G VAJÍČKOVÁ  
16947  Hypermarket  Belgická  MAGGI ASIA 118G INDIE SMAŽ.NUD  
17048  Hypermarket  Belgická                HRANOLKY B" 1KG"  
17251  Hypermarket  Belgická      GRANINI NEKT.1l HRUŠKA 50%  
17900  Hypermarket  Belgická      GRANINI NEKT.1l HRUŠKA 50%

In [146]:
belgicka = df[df["SiteName"] == "Belgická"]
frequent_belgicka = belgicka.groupby("ArticleName")["CostPrice"].agg(lambda x: x.mode()[0] if not x.mode().empty else None)
df.loc[(df["SiteName"] == "Belgická") & (df["CostPrice"].isna()), "CostPrice"] = df["ArticleName"].map(frequent_belgicka)

In [147]:
df[df["SiteName"] == "Belgická"]
test = df[df["CostPrice"].isna()]
test

Empty DataFrame
Columns: [ArticleId, SiteId, ZoneId, Date, Quantity, Price, CostPrice, ZoneName, SiteName, ArticleName]
Index: []

In [148]:
missing_values(df)

ArticleId-0.0%
SiteId-0.0%
ZoneId-0.0%
Date-0.0%
Quantity-0.0%
Price-0.0%
CostPrice-0.0%
ZoneName-0.0%
SiteName-0.0%
ArticleName-0.0%


#### Change data types
- Date change to datetime format
- Quantity change to int format

In [149]:
df["Date"] = pd.to_datetime(df["Date"])

In [150]:
df["Quantity"] = df["Quantity"].astype(int)

In [151]:
df.dtypes

ArticleId               int64
SiteId                  int64
ZoneId                  int64
Date           datetime64[ns]
Quantity                int64
Price                 float64
CostPrice             float64
ZoneName               object
SiteName               object
ArticleName            object
dtype: object

#### Check Prices within ZoneId

In [152]:
check_price = df.groupby(["ZoneId", "ArticleName"])["Price"]
check_price.value_counts()

ZoneId  ArticleName                Price
1       BERLINKI 250G              24.50    485
                                   29.90    172
                                   26.00    126
                                   23.90      7
        BOŽKOV SPE.KÁVOVÝ 30%0,5L  90.50    318
                                           ... 
2       ŠPENÁTOVÝ PROTLAK 400G     58.50     30
                                   54.60     29
                                   40.95     27
                                   38.69      1
                                   41.16      1
Name: count, Length: 221, dtype: int64

In [153]:
def most_frequent(series):
    return series.value_counts().idxmax()

In [154]:
test = df.copy()
test["FrequentPrice"] = check_price.transform(most_frequent)

In [155]:
test["Deviation"] = test["Price"] - test["FrequentPrice"]
test

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.15   
1         171979       7       2 2019-01-18         4   43.5      28.11   
2         200838       5       1 2019-05-16        10   95.0     159.49   
3         176139       7       2 2018-12-21         5  199.0     147.97   
4         172551       5       1 2019-04-11         1   86.7      80.76   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.22   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      77.90   
17994     165811       5       1 2018-12-13         4   21.9      15.74   
17996     170936       8       2 2018-12-05        11   39.0      30.08   

          ZoneName   SiteName                 ArticleName  FrequentPrice  \
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G           39.0   
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ           42.6   
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L           95.0   
3      Supermarket    Pražská            HRANOLKY B" 1KG"          197.0   
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L           90.5   
...            ...        ...                         ...            ...   
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L           89.0   
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G           39.0   
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L           89.0   
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML           21.9   
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G           39.0   

       Deviation  
0            0.0  
1            0.9  
2            0.0  
3            2.0  
4           -3.8  
...          ...  
17991        0.0  
17992        0.0  
17993        0.0  
17994        0.0  
17996        0.0  

[17751 rows x 12 columns]

In [156]:
test["FalsePrice"] = test["Deviation"] != 0

In [157]:
test_dev = test[(test["FalsePrice"] == True) & (test["ZoneId"] == 1) & (test["ArticleId"] == 174022)]
test_dev

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
39        174022       4       1 2019-02-04        -4  259.0     194.49   
46        174022       4       1 2018-12-11         5  229.0     189.51   
75        174022       6       1 2019-02-25         5  266.0     191.07   
78        174022       5       1 2019-01-28         1  259.0     187.06   
87        174022       4       1 2018-11-24         8  259.0     189.83   
...          ...     ...     ...        ...       ...    ...        ...   
17834     174022       6       1 2018-09-22         2  229.0     195.91   
17908     174022       6       1 2019-01-24         6  259.0     190.09   
17909     174022       4       1 2018-11-16         7  259.0     189.83   
17942     174022       6       1 2018-10-09         3  259.0     189.49   
17983     174022       4       1 2018-11-18         3  259.0     189.83   

          ZoneName   SiteName             ArticleName  FrequentPrice  \
39     Hypermarket  Londýnská  JÄGERMEISTER 35%  0,7L          239.0   
46     Hypermarket  Londýnská  JÄGERMEISTER 35%  0,7L          239.0   
75     Hypermarket   Vídeňská  JÄGERMEISTER 35%  0,7L          239.0   
78     Hypermarket   Belgická  JÄGERMEISTER 35%  0,7L          239.0   
87     Hypermarket  Londýnská  JÄGERMEISTER 35%  0,7L          239.0   
...            ...        ...                     ...            ...   
17834  Hypermarket   Vídeňská  JÄGERMEISTER 35%  0,7L          239.0   
17908  Hypermarket   Vídeňská  JÄGERMEISTER 35%  0,7L          239.0   
17909  Hypermarket  Londýnská  JÄGERMEISTER 35%  0,7L          239.0   
17942  Hypermarket   Vídeňská  JÄGERMEISTER 35%  0,7L          239.0   
17983  Hypermarket  Londýnská  JÄGERMEISTER 35%  0,7L          239.0   

       Deviation  FalsePrice  
39          20.0        True  
46         -10.0        True  
75          27.0        True  
78          20.0        True  
87          20.0        True  
...          ...         ...  
17834      -10.0        True  
17908       20.0        True  
17909       20.0        True  
17942       20.0        True  
17983       20.0        True  

[506 rows x 13 columns]

#### Notes
- Apparently, I have a product in one Zone (within one Site location but also outside locations) with different prices. There can be multiple challenges that should be addressed:
    - Price development across time: We should consider whether there is a price change within the date-time period (in this case, sampled over an annual aggregation).
    - Price reduction/discounts: We can consider discount ranges for selected products and determine whether the price is reduced or the base price.
    - Site location can influence price differences within one zone. In this case, SiteName indicates streets. In a real-life scenario, it can indicate regions, and each region may have a different pricing policy.
    - Even within one region or location, the price can differ. Typically, identical articles can have different prices in the city center, where demand is high, compared to the outskirts of the city.
        
**I follow the assumption that each zone should have the same price for a product, so I will substitute the price with the most frequent value for each zone. I originally stored it as FrequentPrice and calculated the deviation from the actual price in the column. I also flagged each row if it deviates from the most frequent price.**

#### Substitute Price with Most Frequent Price for each product and each zone separatelly.

In [158]:
new_price = df.groupby(["ZoneId", "ArticleName"])["Price"]
df["FrequentPrice"] = new_price.transform(most_frequent)
df

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.15   
1         171979       7       2 2019-01-18         4   43.5      28.11   
2         200838       5       1 2019-05-16        10   95.0     159.49   
3         176139       7       2 2018-12-21         5  199.0     147.97   
4         172551       5       1 2019-04-11         1   86.7      80.76   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.22   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      77.90   
17994     165811       5       1 2018-12-13         4   21.9      15.74   
17996     170936       8       2 2018-12-05        11   39.0      30.08   

          ZoneName   SiteName                 ArticleName  FrequentPrice  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G           39.0  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ           42.6  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L           95.0  
3      Supermarket    Pražská            HRANOLKY B" 1KG"          197.0  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L           90.5  
...            ...        ...                         ...            ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L           89.0  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G           39.0  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L           89.0  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML           21.9  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G           39.0  

[17751 rows x 11 columns]

In [159]:
df["Price"] = df["FrequentPrice"]

In [160]:
df.drop(["FrequentPrice"], axis=1, inplace=True)

In [161]:
df

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.15   
1         171979       7       2 2019-01-18         4   42.6      28.11   
2         200838       5       1 2019-05-16        10   95.0     159.49   
3         176139       7       2 2018-12-21         5  197.0     147.97   
4         172551       5       1 2019-04-11         1   90.5      80.76   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.22   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      77.90   
17994     165811       5       1 2018-12-13         4   21.9      15.74   
17996     170936       8       2 2018-12-05        11   39.0      30.08   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17751 rows x 10 columns]

#### Substitute Cost Price with Most Frequent Cost Price for each product and each zone separatelly

In [162]:
checkup = df.copy()
checkup = df[df["CostPrice"] > df["Price"]]
checkup

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
2         200838       5       1 2019-05-16        10   95.0     159.49   
50        170936       4       1 2019-05-16         7   39.0      56.38   
141       172671       5       1 2019-05-15         2   19.9      20.55   
168       165811       6       1 2019-05-17        46   21.9      29.62   
181       200838       7       2 2019-05-18         6   95.0     152.39   
...          ...     ...     ...        ...       ...    ...        ...   
17769     199219       5       1 2019-05-18         9   89.0     146.38   
17776     167108       7       2 2019-05-16         2   26.9      34.82   
17814     170936       8       2 2019-05-17        11   39.0      56.38   
17935     174022       6       1 2019-05-13         4  239.0     383.01   
17953     172551       8       2 2019-05-14        28   90.5     152.33   

          ZoneName   SiteName                 ArticleName  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
50     Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
141    Hypermarket   Belgická    GNOCCHI 40% 500G SPINACI  
168    Hypermarket   Vídeňská  JČ SMETANA KE ŠL.33% 250ML  
181    Supermarket    Pražská     PRAŽSKÁ VODKA 37,5%0,5L  
...            ...        ...                         ...  
17769  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L  
17776  Supermarket    Pražská  GRANINI NEKT.1l HRUŠKA 50%  
17814  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  
17935  Hypermarket   Vídeňská      JÄGERMEISTER 35%  0,7L  
17953  Supermarket   Plzeňská   BOŽKOV SPE.KÁVOVÝ 30%0,5L  

[358 rows x 10 columns]

In [163]:
new_cost = df.groupby(["ZoneId", "ArticleName"])["CostPrice"]
df["FrequentCostPrice"] = new_cost.transform(most_frequent)
df

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.15   
1         171979       7       2 2019-01-18         4   42.6      28.11   
2         200838       5       1 2019-05-16        10   95.0     159.49   
3         176139       7       2 2018-12-21         5  197.0     147.97   
4         172551       5       1 2019-04-11         1   90.5      80.76   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.22   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      77.90   
17994     165811       5       1 2018-12-13         4   21.9      15.74   
17996     170936       8       2 2018-12-05        11   39.0      30.08   

          ZoneName   SiteName                 ArticleName  FrequentCostPrice  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G              30.06  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ              28.11  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L              85.53  
3      Supermarket    Pražská            HRANOLKY B" 1KG"             142.00  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L              81.84  
...            ...        ...                         ...                ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L              78.02  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G              30.06  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L              78.02  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML              15.75  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G              30.06  

[17751 rows x 11 columns]

In [164]:
df["CostPrice"] = df["FrequentCostPrice"]
df.drop(["FrequentCostPrice"], axis=1, inplace=True)
df

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.06   
1         171979       7       2 2019-01-18         4   42.6      28.11   
2         200838       5       1 2019-05-16        10   95.0      85.53   
3         176139       7       2 2018-12-21         5  197.0     142.00   
4         172551       5       1 2019-04-11         1   90.5      81.84   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.02   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      78.02   
17994     165811       5       1 2018-12-13         4   21.9      15.75   
17996     170936       8       2 2018-12-05        11   39.0      30.06   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17751 rows x 10 columns]

#### Notes
- **Why did I substitute the cost price as well?**
    - Cost price had the same issue as Price, so it makes sense to have the same cost price within one zone as well. However, it should be challenged with the same assumptions as Price (basically, what I mentioned in the notes).
    - Before changing the cost price, I also checked whether there was an issue where the cost price was higher than the product price. This can happen during sales or in cases of specific product-selling policies (where a product has a negative margin to increase the number of customers visiting the shop). Typically, hypermarkets reduce prices on certain items to attract customers, who then purchase other products with higher margins. However, in this case, even if I hadn't substituted the Price column with the most frequent price, the deviations were quite extensive. Therefore, substituting the cost price can help address this issue.

#### Quantity below zero

In [218]:
quantity = df.copy()
quantity = quantity[quantity["Quantity"] < 0]
quantity

ArticleId  SiteId  ZoneId        Date  Quantity  Price  CostPrice  \
8         176139       4       1  2019-02-17      -1.0  199.0     142.00   
26        174022       5       1  2019-02-15     -16.0  239.0     196.90   
34        199219       4       1  2019-02-09      -6.0   89.0      78.02   
37        170295       4       1  2019-02-12     -14.0   24.5      15.78   
39        174022       4       1  2019-02-04      -4.0  259.0     194.49   
...          ...     ...     ...         ...       ...    ...        ...   
17852     167108       4       1  2019-02-11      -5.0   27.9      18.56   
17880     165954       5       1  2019-02-12      -3.0    8.7       5.87   
17971     200838       8       2  2019-02-07      -5.0   95.0      81.70   
17982     200838       7       2  2019-02-08      -2.0   95.0      81.29   
17984     172551       7       2  2019-02-02      -1.0   96.9      82.29   

          ZoneName   SiteName                  ArticleName  
8      Hypermarket  Londýnská             HRANOLKY B" 1KG"  
26     Hypermarket   Belgická       JÄGERMEISTER 35%  0,7L  
34     Hypermarket  Londýnská    NORDIC ICE VODKA37,5%0,5L  
37     Hypermarket  Londýnská                BERLINKI 250G  
39     Hypermarket  Londýnská       JÄGERMEISTER 35%  0,7L  
...            ...        ...                          ...  
17852  Hypermarket  Londýnská   GRANINI NEKT.1l HRUŠKA 50%  
17880  Hypermarket   Belgická  LIPÁNEK UHT MLÉKO 250ML KAK  
17971  Supermarket   Plzeňská      PRAŽSKÁ VODKA 37,5%0,5L  
17982  Supermarket    Pražská      PRAŽSKÁ VODKA 37,5%0,5L  
17984  Supermarket    Pražská    BOŽKOV SPE.KÁVOVÝ 30%0,5L  

[820 rows x 10 columns]

In [170]:
df["Quantity"] = df["Quantity"].abs()
df

ArticleId  SiteId  ZoneId       Date  Quantity  Price  CostPrice  \
0         170936       4       1 2019-05-09         7   39.0      30.06   
1         171979       7       2 2019-01-18         4   42.6      28.11   
2         200838       5       1 2019-05-16        10   95.0      85.53   
3         176139       7       2 2018-12-21         5  197.0     142.00   
4         172551       5       1 2019-04-11         1   90.5      81.84   
...          ...     ...     ...        ...       ...    ...        ...   
17991     199219       5       1 2019-05-04        17   89.0      78.02   
17992     170936       6       1 2018-09-10         5   39.0      30.06   
17993     199219       6       1 2019-01-23         4   89.0      78.02   
17994     165811       5       1 2018-12-13         4   21.9      15.75   
17996     170936       8       2 2018-12-05        11   39.0      30.06   

          ZoneName   SiteName                 ArticleName  
0      Hypermarket  Londýnská      ŠPENÁTOVÝ PROTLAK 400G  
1      Supermarket    Pražská    POMAZÁNKA 120G VAJÍČKOVÁ  
2      Hypermarket   Belgická     PRAŽSKÁ VODKA 37,5%0,5L  
3      Supermarket    Pražská            HRANOLKY B" 1KG"  
4      Hypermarket   Belgická   BOŽKOV SPE.KÁVOVÝ 30%0,5L  
...            ...        ...                         ...  
17991  Hypermarket   Belgická   NORDIC ICE VODKA37,5%0,5L  
17992  Hypermarket   Vídeňská      ŠPENÁTOVÝ PROTLAK 400G  
17993  Hypermarket   Vídeňská   NORDIC ICE VODKA37,5%0,5L  
17994  Hypermarket   Belgická  JČ SMETANA KE ŠL.33% 250ML  
17996  Supermarket   Plzeňská      ŠPENÁTOVÝ PROTLAK 400G  

[17751 rows x 10 columns]

#### Notes
- In this case, I'm not sure how to resolve this issue based on the assumptions. Possible explanations:
    - Negative quantity represents a returned product (low confidence: why would it be in sales?).
    - Negative quantity represents a giveaway item (low confidence: sometimes the quantity is too high to be a giveaway).
    - Negative quantity represents a defective/discontinued product or an item past its expiration date (mid confidence: why would it be in sales?).
    - Negative quantity is a typo (low confidence: there are too many rows for this to be likely).
    - Negative quantity is the result of faulty extraction, import, or processing (high confidence: since this was not addressed in the assumptions, I can consider it faulty data).

**I would challenge the consultant or data provider or check the extraction process to identify the root cause of this issue. For now, I will simply change the negative quantity values to positive numbers. If this approach turns out to be incorrect, I will adjust the solution accordingly.**

In [171]:
value_count(df)

ArticleId
170936    1826
174022    1774
200838    1652
172551    1612
167108    1606
199219    1580
165811    1550
170295    1171
176139    1055
171979    1000
194516     951
170069     644
169281     533
172671     497
165954     300
Name: count, dtype: int64


SiteId
4    3741
5    3720
6    3443
7    3379
8    3215
9     253
Name: count, dtype: int64


ZoneId
1    10904
2     6847
Name: count, dtype: int64


Date
2018-11-15    69
2018-11-16    68
2018-11-14    68
2018-11-23    67
2018-11-24    65
              ..
2018-12-06    33
2018-12-10    33
2018-08-06    31
2018-07-30    28
2018-12-24    21
Name: count, Length: 389, dtype: int64


Quantity
1      3134
2      2372
3      1782
4      1531
5      1259
       ... 
273       1
432       1
62        1
56        1
76        1
Name: count, Length: 68, dtype: int64


Price
21.9     2083
39.0     1826
239.0    1774
95.0     1652
90.5     1612
26.9     1606
89.0     1580
24.5     1171
42.6     1000
43.8      951
59.9      644
197.0     5

In [173]:
df[["Quantity", "Price", "CostPrice"]].describe()

Quantity         Price     CostPrice
count  17751.000000  17751.000000  17751.000000
mean       6.831953     79.327491     62.434946
std        7.996738     69.863429     55.479868
min        1.000000      8.700000      5.870000
25%        2.000000     26.900000     18.580000
50%        5.000000     43.800000     30.060000
75%        9.000000     95.000000     82.290000
max      432.000000    239.000000    191.900000